### PR TITLE
Scoring and discussion functionality

### PR DESCRIPTION
**Pull request by NeonBohdan**
# Description
Adds `fastchat_score_input` and `fastchat_discussion_input` queues to process vote and discution requests from klat

# Issues

# Other Notes

### CODE REVIEW

File: neon_llm_fastchat/rmq.py [hunk 1]
```diff
@@ -93,3 +111,81 @@ def handle_request(self,
...
+        prompt = f'Why Answer "{best_responce}" to the Question "{query}" generated by Bot named "{bot_name}" is good?'
+
+        opinion = self.fastChat.ask(message=prompt, chat_history=[])
+
+        api_response = {
+            "message_id": message_id,
+            "opinion": opinion
+        }
+
+        channel.basic_publish(exchange='',
```
Comment (NeonKirill) [inline neon_llm_fastchat/rmq.py]: please use built-in `send_message` method

File: neon_llm_fastchat/fastchat.py [hunk 2]
```diff
@@ -78,3 +86,33 @@ def call_model(self, prompt):
...
+        results = self.model.score_batch(
+            source = tokens_list,
+            target = target_tokens_list,
+        )
+
+        log_probs_list = [result.log_probs for result in results]
+        return log_probs_list
+
+    @staticmethod
+    def compute_ppl(log_probs):
```
Comment (NeonKirill) [inline neon_llm_fastchat/fastchat.py]: please insert ppl formula for the future reference

File: neon_llm_fastchat/rmq.py [hunk 3]
```diff
@@ -60,6 +60,24 @@ def __init__(self):
                                on_error=self.default_error_handler,
                                auto_ack=False)
 
+        self.queue_score = "fastchat_score_input"
```
Comment (NeonKirill) [inline neon_llm_fastchat/rmq.py]: please make it as `properties`, I think it worth to make single interface for LLM's in future, and having same common components will be useful there..

File: neon_llm_fastchat/rmq.py [hunk 4]
```diff
@@ -93,3 +111,81 @@ def handle_request(self,
...
+        :param channel: MQ channel object (pika.channel.Channel)
+        :param method: MQ return method (pika.spec.Basic.Return)
+        :param body: request body (dict)
+        """
+        message_id = body["message_id"]
+        routing_key = body["routing_key"]
+
+        query = body["query"]
+        options = body["options"]
+        responses = list(options.values())
```
Comment (NeonKirill) [inline neon_llm_fastchat/rmq.py:159]: handle edge-case where options are empty

File: neon_llm_fastchat/fastchat.py [hunk 5]
```diff
@@ -78,3 +86,33 @@ def call_model(self, prompt):
...
+        ppl = np.exp(-np.mean(log_probs))
+        return ppl
+
+    def ppl(self, question, answers):
```
Comment (NeonKirill) [inline neon_llm_fastchat/fastchat.py]: pls add type annotations

File: neon_llm_fastchat/fastchat.py [hunk 6]
```diff
@@ -78,3 +86,33 @@ def call_model(self, prompt):
         output_tokens = results[0].hypotheses[0]
         text = self.tokenizer.decode(self.tokenizer.convert_tokens_to_ids(output_tokens), spaces_between_special_tokens=False)
         return text
+
+    def call_score(self, prompt, targets):
```
Comment (NeonKirill) [inline neon_llm_fastchat/fastchat.py]: pls add type annotations

File: neon_llm_fastchat/fastchat.py [hunk 7]
```diff
@@ -54,19 +55,26 @@ def convert_role(role):
...
         prompt = self.system_message
         # Context N messages
         for role, content in chat_history[-self.context_depth:]:
             role_fastchat = self.convert_role(role)
             prompt += f"### {role_fastchat}: {content}\n"
         prompt += f"### Human: {message}\n### Assistant:"
-        
+        return prompt
+
+    def ask(self, message, chat_history):
```
Comment (NeonKirill) [inline neon_llm_fastchat/fastchat.py]: pls add type annotations // not clear what this method is doing, so consider either renaming or adding docstring here

File: neon_llm_fastchat/fastchat.py [hunk 8]
```diff
@@ -54,19 +55,26 @@ def convert_role(role):
...
+        prompt = self.assemble_prompt(message, chat_history)
         bot_message = self.call_model(prompt)
         return bot_message
 
-    def call_model(self, prompt):
+    def tokenize(self, prompt):
```
Comment (NeonKirill) [inline neon_llm_fastchat/fastchat.py]: pls add type annotations including return types, as it is not clear what we expect

**Review by NeonKirill — CHANGES_REQUESTED** [review]

File: neon_llm_fastchat/rmq.py (hunk 3 above)
Comment (NeonBohdan) [reply neon_llm_fastchat/rmq.py]: Done with https://github.com/NeonGeckoCom/neon-llm-fastchat/pull/5/commits/220c66330066f364676e0017043b556f7c61b7fd

File: neon_llm_fastchat/fastchat.py (hunk 5 above)
Comment (NeonBohdan) [reply neon_llm_fastchat/fastchat.py]: Done with https://github.com/NeonGeckoCom/neon-llm-fastchat/pull/5/commits/b5901b5390a0f16ad9b54d940c1db952f6a71fae

File: neon_llm_fastchat/fastchat.py (hunk 6 above)
Comment (NeonBohdan) [reply neon_llm_fastchat/fastchat.py]: Done with https://github.com/NeonGeckoCom/neon-llm-fastchat/pull/5/commits/b5901b5390a0f16ad9b54d940c1db952f6a71fae

File: neon_llm_fastchat/fastchat.py (hunk 8 above)
Comment (NeonBohdan) [reply neon_llm_fastchat/fastchat.py]: Done with https://github.com/NeonGeckoCom/neon-llm-fastchat/pull/5/commits/b5901b5390a0f16ad9b54d940c1db952f6a71fae

File: neon_llm_fastchat/rmq.py (hunk 1 above)
Comment (NeonBohdan) [reply neon_llm_fastchat/rmq.py]: `send_message` only supports `auto_ack`

File: neon_llm_fastchat/rmq.py (hunk 4 above)
Comment (NeonBohdan) [reply neon_llm_fastchat/rmq.py:159]: Done with https://github.com/NeonGeckoCom/neon-llm-fastchat/pull/5/commits/8a7067d8c7f3b0371265096e2dcff65f3496443b

File: neon_llm_fastchat/fastchat.py (hunk 2 above)
Comment (NeonBohdan) [reply neon_llm_fastchat/fastchat.py]: Done with https://github.com/NeonGeckoCom/neon-llm-fastchat/pull/5/commits/b4ab71a7ac20b2de693a7798cc8e318385c88c39

File: neon_llm_fastchat/fastchat.py (hunk 7 above)
Comment (NeonBohdan) [reply neon_llm_fastchat/fastchat.py]: Done with https://github.com/NeonGeckoCom/neon-llm-fastchat/pull/5/commits/b4ab71a7ac20b2de693a7798cc8e318385c88c39

File: neon_llm_fastchat/fastchat.py [hunk 9]
```diff
@@ -78,3 +89,34 @@ def call_model(self, prompt):
...
+        return ppl
+
+    def ppl(self, question: str, answers: List[str]) -> List[float]:
+        question_prompt = self.assemble_prompt(question, [])
+        log_probs_list = self.call_score(question_prompt, answers)
+        ppl_list = [self.compute_ppl(log_probs) for log_probs in log_probs_list]
+        return ppl_list
+    
+    @staticmethod
+    def get_best(ppl_list: List[float]) -> int:
```
Comment (NeonDaniel) [inline neon_llm_fastchat/fastchat.py]: Could you add docstrings and describe the parameters/return for these methods; I'm not sure what the args/return values represent for some of these

**Review by NeonDaniel — APPROVED** [review]

**Review by NeonKirill — APPROVED** [review]

**Pull request merged**